### PR TITLE
New error message format

### PR DESCRIPTION
**Pull request by sonalmahajan15**
This PR introduces a new, improved error message format with the aim of improving UX.

New message format:

```
Potential nil panic detected. Observed nil flow from source to dereference point:
	-> CONSUMER_POS: explanation text from producer and consumer
	-> CONSUMER_POS: explanation text from producer and consumer
        ...
	-> CONSUMER_POS: explanation text from producer and consumer
```

`conflict.go` is revamped to allow conformance to the new error message format. For simplicity of this PR, the grouping logic is removed. It will be added back in a follow up PR.

### CODE REVIEW

File: assertion/function/assertiontree/preprocess_blocks.go
Comment (sonalmahajan15) [inline assertion/function/assertiontree/preprocess_blocks.go:90]: Fix applied for a NilAway reported error, which seems to be a valid one. Not sure why it did not get reported earlier, though.

File: inference/engine.go
Comment (sonalmahajan15) [inline inference/engine.go:409]: The structs `TrueBecauseMyopia` and `FalseBecauseMyopia` and their methods are removed because they were unused.

File: inference/conflict.go
Comment (lazaroclapp) [inline inference/conflict.go:56]: Is this one of those cases where a method supports nilable receiver? (since I see from the function above that `n` could just be `node{}`)

File: inference/conflict.go
Comment (lazaroclapp) [inline inference/conflict.go:67]: Feels like this doc should be in the `newNode` function now. No?

File: inference/conflict.go
Comment (lazaroclapp) [inline inference/conflict.go:71]: This is missing some info, which I think is that these nodes get added starting from the conflict node, so they are added in reverse order from what the program flow would look like.

File: inference/conflict.go
Comment (lazaroclapp) [inline inference/conflict.go:138]: Can I get a one sentence reminder in comments here about what these two cases for `t` represent? :)

File: inference/explained_bool.go
Comment (lazaroclapp) [inline inference/explained_bool.go:182]: Weird that this line isn't reached by tests, actually (not something to fix in this PR but it strikes me as curious. When is `FalseBecauseAnnotation` used?)

File: testdata/src/go.uber.org/anonymousfunction/anonymous_functions_assignment.go
Comment (lazaroclapp) [inline testdata/src/go.uber.org/anonymousfunction/anonymous_functions_assignment.go:22]: Interesting. What is causing us to switch these to get reported here vs at lambda declaration? (This is better, but I wonder if it's an intentional change or some consequence of the new error reporting method)

Edit: Just realized that the code above keeps updating the conflict position for as long as it finds nodes in its non-nil path, so the reporting position of the conflict ends up being that of the final non-nil consumer in the flow, which is what we want. Nice! (Maybe worth a once sentence comment in the code where this happens)

**Review by lazaroclapp — COMMENTED** [review]
This is fantastic 🚀 ! Looks good to me, but I had a few questions / requests for more comments, so I'll go ahead and approve it once those are answered.

Btw, I looked at the actual code in detail, but the changes to testdata are a lot, so I mostly just eyeballed them. If you think there might be issues there, I can go line by line 😅

File: inference/conflict.go
Comment (sonalmahajan15) [reply inference/conflict.go:56]: Not really. `newNode` is returning a struct, not a pointer, so the receiver `n` here of `String()` is never nil. Also, `n.position`, which is of type `token.Position`, is also a struct, not a pointer, so we can safely invoke the method `IsValid()` on it.

File: inference/conflict.go
Comment (sonalmahajan15) [reply inference/conflict.go:67]: Ah, good catch. Moved it to `newNode`.

File: inference/conflict.go
Comment (sonalmahajan15) [reply inference/conflict.go:71]: Your are right. Modified the comment to explain this better.

File: inference/conflict.go
Comment (sonalmahajan15) [reply inference/conflict.go:138]: Yes, added a comment :)

File: inference/explained_bool.go
Comment (sonalmahajan15) [reply inference/explained_bool.go:182]: Hmm.. `FalseBecauseAnnotation` is used when a program site is explicitly annotated as `nonnil`. I can see a couple of instances in our test cases, where this is present. So, not sure why isn't it covered by tests. I'll look into it.

File: inference/conflict.go
Comment (lazaroclapp) [reply inference/conflict.go:56]: Ah, right, too used to Java reference types. I did mean `n.position` would be `nil` by default, but, you are right, since it is a struct type and not a pointer to struct, it will just be an empty struct itself...

File: testdata/src/go.uber.org/anonymousfunction/anonymous_functions_assignment.go
Comment (sonalmahajan15) [reply testdata/src/go.uber.org/anonymousfunction/anonymous_functions_assignment.go:22]: Exactly, you got the reasoning right. Added a comment explaining this idea.

**Comment by sonalmahajan15** [comment]
> This is fantastic 🚀 ! Looks good to me, but I had a few questions / requests for more comments, so I'll go ahead and approve it once those are answered.
> 
> Btw, I looked at the actual code in detail, but the changes to testdata are a lot, so I mostly just eyeballed them. If you think there might be issues there, I can go line by line 😅

Thanks @lazaroclapp. I have addressed / answered your comments. Please feel free to do a second pass.

I know, the test cases do have a lot of changes -- they are either about moving the position of where the error gets reported or modifying the text of `want` messages to suit the new format. I have ensured that there is nothing unexpected and the test cases pass as well, so feel free to not inspect them closely. :)

**Review by lazaroclapp — APPROVED** [review]
LGTM! (Though I still have my questions about that coverage issue for the `FalseBecauseAnnotation` explained bool...)

**Comment by sonalmahajan15** [comment]
> LGTM! (Though I still have my questions about that coverage issue for the `FalseBecauseAnnotation` explained bool...)

I checked this issue and turns out that we did not have a viable test case for covering `FalseBecauseAnnotation` in our test suite, particularly in the full infer mode. I have added a test case in `testdata/src/go.uber.org/inference/inference.go` for the same, which now shows complete coverage of `FalseBecauseAnnotation`.

File: inference/conflict.go
Comment (sonalmahajan15) [reply inference/conflict.go:26]: I like this. I'll create a refactoring diff at the top of this PR stack.

File: inference/explained_bool.go
Comment (yuxincs) [inline inference/explained_bool.go:34]:
```suggestion
	DeeperReason() ExplainedBool
```
Right? because `ExplainedBool.getExplainedBool` reads a little weird to me ;)

File: util/util.go
Comment (yuxincs) [inline util/util.go:479]: Is it possible that we integrate (in the future of course) this pretty print logic into the diagnostics engine organically? 

Post-process via regex always feels a bit hacky to me :)

File: inference/conflict.go
Comment (sonalmahajan15) [reply inference/conflict.go:27]: Fair point. Renamed and added comment.

File: inference/conflict.go
Comment (sonalmahajan15) [reply inference/conflict.go:28]: Done

File: util/util.go
Comment (sonalmahajan15) [reply util/util.go:479]: Agreed, should be doable. I'll try and incorporate this in the planned refactor.

File: inference/conflict.go
Comment (sonalmahajan15) [reply inference/conflict.go:161]: Agreed. Done

File: inference/explained_bool.go
Comment (sonalmahajan15) [reply inference/explained_bool.go:34]: Good suggestion. I just un-exported `deeperReason()` since it is only used locally from within the pakcage.

File: inference/conflict.go
Comment (sonalmahajan15) [reply inference/conflict.go:81]: That's a great suggestion. Addressed :)

File: inference/conflict.go
Comment (sonalmahajan15) [reply inference/conflict.go:91]: > But I think this does not add too much overhead anyway.
That is true, it is quite inexpensive, but might as well do it the right way. Added a TODO to act upon later.

File: inference/conflict.go
Comment (sonalmahajan15) [reply inference/conflict.go:114]: True. I'll prefer doing a refactoring at the top of the stack, or as a follow-up after this existing stack lands.

**Pull request merged**
